File: leetcode-daily/april-2024/85-maximal-rectangle.cpp
/*
https://leetcode.com/problems/maximal-rectangle

Given a `rows x cols` binary `matrix` filled with `0`'s and `1`'s, find the largest rectangle
containing only `1`'s and return *its area*.


**Example 1:**

**Input:** matrix =
[["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
**Output:** 6
**Explanation:** The maximal rectangle is shown in the above picture.

**Example 2:**

**Input:** matrix = [["0"]]
**Output:** 0

**Example 3:**

**Input:** matrix = [["1"]]
**Output:** 1


**Constraints:**

* `rows == matrix.length`
* `cols == matrix[i].length`
* `1 <= row, cols <= 200`
* `matrix[i][j]` is `'0'` or `'1'`.
*/

#include "./../../cpp-utils/printutils.h"
#include "bits/stdc++.h"
#include <iostream>
using namespace std;

class Solution {
  private:
    int largestRectangleArea(vector<int> &heights) {
        stack<pair<int, int>> st;
        int i = 0, n = heights.size(), max_area = 0, begin = 0;
        while (i < n) {
            begin = i;
            while (!st.empty() && heights[i] < st.top().second) {
                max_area = max(max_area, (i - st.top().first) * st.top().second);
                begin = st.top().first;
                st.pop();
            }
            st.push({begin, heights[i]});
            i++;
        }

        while (!st.empty()) {
            max_area = max(max_area, (i - st.top().first) * st.top().second);
            st.pop();
        }
        return max_area;
    }

  public:
    int maximalRectangle(vector<vector<char>> &matrix) {
        int m = matrix.size();
        int n = matrix[0].size();

        vector<vector<int>> heights(m, vector<int>(n, 0));
        for (int j = 0; j < n; ++j) {
            for (int i = 0; i < m; ++i) {
                int height = 0;
                if (matrix[i][j] == '1') {
                    if (i > 0 && heights[i - 1][j] > 1) {
                        height = heights[i - 1][j] - 1;
                    } else {
                        height = 0;
                        int k = i;
                        while (k < m && matrix[k][j] == '1') {
                            height++;
                            k++;
                        }
                    }
                }
                heights[i][j] = height;
            }
        }

        int max_area = INT_MIN;
        for (int i = 0; i < m; ++i) {
            max_area = max(max_area, largestRectangleArea(heights[i]));
        }
        return max_area;
    }
};

int main() {
    Solution sol;

    vector<vector<char>> matrix_1 = {{'1', '0', '1', '0', '0'},
                                     {'1', '0', '1', '1', '1'},
                                     {'1', '1', '1', '1', '1'},
                                     {'1', '0', '0', '1', '0'}};
    int expected_1 = 6;
    int output_1 = sol.maximalRectangle(matrix_1);
    cout << output_1 << endl;
    // assert(output_1 == expected_1);

    vector<vector<char>> matrix_2 = {{'0'}};
    int expected_2 = 0;
    int output_2 = sol.maximalRectangle(matrix_2);
    cout << output_2 << endl;
    // assert(output_2 == expected_2);

    vector<vector<char>> matrix_3 = {{'1'}};
    int expected_3 = 1;
    int output_3 = sol.maximalRectangle(matrix_3);
    cout << output_3 << endl;
    // assert(output_3 == expected_3);

    vector<vector<char>> matrix_4 = {
        {'0', '0', '1', '0'}, {'0', '0', '1', '0'}, {'0', '0', '1', '0'}, {'0', '0', '1', '1'},
        {'0', '1', '1', '1'}, {'0', '1', '1', '1'}, {'1', '1', '1', '1'}};
    int expected_4 = 9;
    int output_4 = sol.maximalRectangle(matrix_4);
    cout << output_4 << endl;
    // assert(output_4 == expected_4);

    vector<vector<char>> matrix_5 = {{'0', '0', '1'}, {'1', '1', '1'}};
    int expected_5 = 3;
    int output_5 = sol.maximalRectangle(matrix_5);
    cout << output_5 << endl;
    // assert(output_5 == expected_5);

    vector<vector<char>> matrix_6 = {
        {'1', '0', '0', '0', '1'}, {'1', '1', '0', '1', '1'}, {'1', '1', '1', '1', '1'}};
    int expected_6 = 5;
    int output_6 = sol.maximalRectangle(matrix_6);
    cout << output_6 << endl;
    // assert(output_6 == expected_6);

    return 0;
}

File: leetcode-daily/april-2024/85-maximal-rectangle.rs
/*
https://leetcode.com/problems/maximal-rectangle

Given a `rows x cols` binary `matrix` filled with `0`'s and `1`'s, find the largest rectangle containing only `1`'s and
return *its area*.


**Example 1:**

**Input:** matrix = [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
**Output:** 6
**Explanation:** The maximal rectangle is shown in the above picture.

**Example 2:**

**Input:** matrix = [["0"]]
**Output:** 0

**Example 3:**

**Input:** matrix = [["1"]]
**Output:** 1


**Constraints:**

* `rows == matrix.length`
* `cols == matrix[i].length`
* `1 <= row, cols <= 200`
* `matrix[i][j]` is `'0'` or `'1'`.
*/

struct Solution;
impl Solution {
    fn largest_rectangle_area(heights: &Vec<i32>) -> i32 {
        let n: usize = heights.len();
        let mut i: usize = 0;
        let mut begin: usize;
        let mut st: Vec<(usize, i32)> = Vec::new();
        let mut max_area: i32 = 0;

        while i < n {
            begin = i;
            while !st.is_empty() && heights[i] < st[st.len() - 1].1 {
                max_area = i32::max(
                    max_area,
                    (i - st[st.len() - 1].0) as i32 * st[st.len() - 1].1,
                );
                begin = st[st.len() - 1].0;
                st.pop();
            }
            st.push((begin, heights[i]));
            i += 1;
        }

        while !st.is_empty() {
            max_area = i32::max(
                max_area,
                (i - st[st.len() - 1].0) as i32 * st[st.len() - 1].1,
            );
            begin = st[st.len() - 1].0;
            st.pop();
        }

        max_area
    }

    pub fn maximal_rectangle(matrix: Vec<Vec<char>>) -> i32 {
        let m: usize = matrix.len();
        let n: usize = matrix[0].len();
        let mut heights: Vec<Vec<i32>> = vec![vec![0; n]; m];
        let mut h: i32;

        for j in 0..n {
            for i in 0..m {
                h = 0;
                if matrix[i][j] == '1' {
                    if i > 0 && heights[i - 1][j] > 1 {
                        h = heights[i - 1][j] - 1;
                    } else {
                        let mut k: usize = i;
                        while k < m && matrix[k][j] == '1' {
                            h += 1;
                            k += 1;
                        }
                    }
                }
                heights[i][j] = h;
            }
        }

        let mut max_area: i32 = 0;
        for i in 0..m {
            max_area =
                i32::max(max_area, Self::largest_rectangle_area(&heights[i]));
        }

        max_area
    }
}

fn main() {
    let matrix_1: Vec<Vec<char>> = vec![
        vec!['1', '0', '1', '0', '0'],
        vec!['1', '0', '1', '1', '1'],
        vec!['1', '1', '1', '1', '1'],
        vec!['1', '0', '0', '1', '0'],
    ];
    let expected_1: i32 = 6;
    let output_1: i32 = Solution::maximal_rectangle(matrix_1);
    println!("{:?}", output_1);
    assert!(output_1 == expected_1);

    let matrix_2: Vec<Vec<char>> = vec![vec!['0']];
    let expected_2: i32 = 0;
    let output_2: i32 = Solution::maximal_rectangle(matrix_2);
    println!("{:?}", output_2);
    assert!(output_2 == expected_2);

    let matrix_3: Vec<Vec<char>> = vec![vec!['1']];
    let expected_3: i32 = 1;
    let output_3: i32 = Solution::maximal_rectangle(matrix_3);
    println!("{:?}", output_3);
    assert!(output_3 == expected_3);
}
